### PR TITLE
alter_table: Durable Catalog Migration

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR re-keys everything in the durable Catalog on `CatalogItemId` instead of `GlobalId`, and shims everything above the durable Catalog objects using two new methods `GlobalId::to_item_id()`, and `CatalogItemId::to_global_id()`.

Now every item in the Catalog has the following fields:

* `id: CatalogItemId`, a stable _external_ identifier (i.e. in Catalog tables like `mz_objects`) that is the same for the entire lifetime of the object.
* `global_id: GlobalId`, a stable _internal_ identifier for this object that can be used by storage and compute.
* `extra_versions: BTreeMap<Version, GlobalId>`, mapping of versions of an object to the `GlobalId`s used by compute and storage to refer to a specific version.

This de-coupling of `CatalogItemId` and `GlobalId` achieves two things:

1. Externally objects have a stable identifier, even as they are `ALTER`-ed. This is required for external tools like dbt and Terraform that track objects by ID.
2. Internally a `GlobalId` always refers to the same `RelationDesc` + Persist Shard, this maintains the concept from the formalism that a `GlobalId` is never re-assigned to a new pTVC.

The implementation of `ALTER TABLE ... ADD COLUMN ...` will thus allocate a new `GlobalId` which will immutably refer to that specific version of the table.

#### Other Changes

Along with `ItemKey` and `ItemValue` I updated the following Catalog types:

* `GidMappingValue`: replaced the `id` field with `catalog_id` and `global_id`, used to identify builtin catalog objects.
* `ClusterIntrospectionSourceIndexValue`: replaced the `index_id` field with `catalog_id` and `global_id`, used to identify builtin introspection source indexes.
* `CommentKey`: replaced `GlobalId` with `CatalogItemId`, used to identify comments on objects.
* `SourceReferencesKey`: replaced `GlobalId` with `CatalogItemId`, used to track references between a Source and the subsources/tables that read from it.

#### Partial Progress

Today `CatalogItemId` is 1:1 with `GlobalId`, this allows us to implement the `to_item_id` and `to_global_id` shim methods. Until we support `ALTER TABLE ... ADD COLUMN ...` we can freely convert between the two. This allows us to break this change up among multiple PRs instead of a single massive change.

#### Initial Migration

Because `CatalogItemId` and `GlobalId` are currently 1:1, this allows us to migrate the raw values of the IDs, e.g. `GlobalId::User(42)` becomes `CatalogItemId::User(42)`, which is exactly what this PR does.

### Motivation

Progress towards https://github.com/MaterializeInc/database-issues/issues/8233
Implements changes described in https://github.com/MaterializeInc/materialize/pull/30019

### Tips for reviewer

This PR is split into two commits:

1. The durable catalog migration, and updates to durable Catalog objects. I would appreciate the most thorough reviews on this commit.
4. Shimming all calling code to convert between `CatalogItemId` and `GlobalId`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/sql/src/names.rs [hunk 1]
```diff
@@ -1137,17 +1137,17 @@ impl From<&GlobalId> for ObjectId {
...
+            | CommentObjectId::View(item_id)
+            | CommentObjectId::MaterializedView(item_id)
+            | CommentObjectId::Source(item_id)
+            | CommentObjectId::Sink(item_id)
+            | CommentObjectId::Index(item_id)
+            | CommentObjectId::Func(item_id)
+            | CommentObjectId::Connection(item_id)
+            | CommentObjectId::Type(item_id)
+            | CommentObjectId::Secret(item_id)
+            | CommentObjectId::ContinualTask(item_id) => ObjectId::Item(item_id.to_global_id()),
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/sql/src/names.rs:1150]: Totally! That switch over will be in the next PR

File: src/catalog/src/durable/objects.rs [hunk 2]
```diff
@@ -457,13 +454,15 @@ impl From<mz_controller::clusters::ReplicaLocation> for ReplicaLocation {
...
-    pub id: GlobalId,
+    pub id: CatalogItemId,
     pub oid: u32,
+    pub global_id: GlobalId,
     pub schema_id: SchemaId,
     pub name: String,
     pub create_sql: String,
     pub owner_id: RoleId,
     pub privileges: Vec<MzAclItem>,
+    pub extra_versions: BTreeMap<RelationVersion, GlobalId>,
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/catalog/src/durable/objects.rs:465]: > Will we have a single Item entry in the durable catalog per CatalogItemId or a single entry per GlobalId?

There will be a single `Item` entry in the durable catalog per `CatalogItemId`. Each `Item` can then also be optionally referred to by multiple `GlobalId`s, where each `GlobalId` refers to a specific version of the `Item`.

[the remark this answers is not in the document]
Comment (jkosh44) [reply src/catalog/src/durable/objects.rs:465]: So the `global_id` field will be the most recent `GlobalId` and everything else will be in `extra_versions`? Or will the most recent `GlobalId` also be in `extra_versions`?

[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/catalog/src/durable/objects.rs:465]: Great question! Right now what I was thinking is the original `GlobalId` will be in the `global_id` field, i.e. `Version(0)`. `Version(1)` and above is what goes into the `extra_versions` map

**Pull request merged**